File: Programming Pragmatics/C/genericprog.c
/*
 * Mike Zeng
 * CECS 424-05
 * Assignment 3
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    char* key;
    int value;
} person;
//defining compares to sort
int dcompare(const void *a,const void *b) {
    double *x = (double *) a;
    double *y = (double *) b;
    if (*x < *y) return -1;
    else if (*x > *y) return 1;
    return 0;
}
int pcomparefst(const void *a,const void *b) {
    person *x = (person *) a;
    person *y = (person *) b;
    return strcmp(x->key, y->key);
}
int pcomparesnd(const void *a,const void *b) {
    person *x = (person *) a;
    person *y = (person *) b;
    if((x->value - y->value) == 0)  //if their age is the same, go alphabetically
        return strcmp(x->key, y->key);
    else                            //order by descending age
        return y->value - x->value;
}

//data structures
char *names[] = {"Hal","Susann","Dwight","Kassandra","Lawrence","Cindy","Cory","Mac","Romana","Doretha","Danna","Zara","Rosalyn","Risa","Benny","Juan","Natalie"};
int ages[] = {20, 31, 19, 21, 25, 22, 27, 19, 27, 32, 20, 23, 26, 24, 28, 33, 25};
double numbers[] = { 645.32, 37.40, 76.30, 5.40, -34.23, 1.11, -34.94, 23.37, 635.46, -876.22, 467.73, 62.26 };
const int num_len = (sizeof numbers/ sizeof *numbers);
const int people_len = (sizeof names/ sizeof *names);

//display lists
void display_num(double* a){
    for(int i=0;i<num_len;i++){
        printf("%.2f  ", a[i]);
    }
}
void display_people(person* p){
    for(int i=0;i<people_len;i++){
        printf("%s,%d ", p[i].key, p[i].value);
    }
}

int main(int argc, char** argv) {   
    person *people = malloc(sizeof(person) * people_len);
    if (people == NULL) {   //people error check
        perror("Malloc");
        exit(EXIT_FAILURE);
    }
    for(int i = 0;i<people_len;i++){    //generate the people list
        people[i].key = names[i];
        people[i].value = ages[i];
    }
    
    //display list of numbers
    printf("List of numbers: \n");
    display_num(numbers);
    qsort(numbers, num_len, sizeof(*numbers), dcompare);    //generic qsort to sort in ascending order
    printf("\nOrdered numbers: \n");
    display_num(numbers);
    
    //display list of people
    printf("\n\nList of people: \n");
    display_people(people);
    person *alpha_sort_people = malloc(sizeof(person) * people_len);
    for(int i = 0;i<people_len;i++){    //generate temp people list to sort alphabetically
        alpha_sort_people[i].key = people[i].key;
        alpha_sort_people[i].value = people[i].value;
    }
    qsort(alpha_sort_people, people_len, sizeof(*alpha_sort_people), pcomparefst);  //alphabetically sorted people
    printf("\nOrdered people by name alphabetically:\n");
    display_people(alpha_sort_people);
    qsort(alpha_sort_people, people_len, sizeof(*alpha_sort_people), pcomparesnd);  //age descended sort people, then by alphabetically
    printf("\nOrdered people by descending age then alphabetically:\n");
    display_people(alpha_sort_people);
    printf("\n");        
    
    free(people);
    free(alpha_sort_people);
    return (EXIT_SUCCESS);
}

File: Programming Pragmatics/C#/genericprog.cs
/*
 * Mike Zeng
 * CECS 424-05
 * Assignment 3
 */
using System;
using System.Collections.Generic;
using System.Linq;

namespace _424hw3sharp
{
    class Assignment3
    {
        //display functions
        static void display_num(List<double> d){
            for (int i = 0; i < d.Count; i++)
                Console.Write(d[i]+" ");
        }

        static void display_people(Dictionary<string, int> p){
            foreach (KeyValuePair<string, int> person in p)
                Console.Write(person.Key + ","+ person.Value + " ");
        }

        static void Main(string[] args)
        {
            //data structures
            List<double> numbers = new List<double> { 645.32, 37.40, 76.30, 5.40, -34.23, 1.11, -34.94, 23.37, 635.46, -876.22, 467.73, 62.26 };
            Dictionary<string, int> people = new Dictionary<string, int>() { { "Hal", 20 }, {"Susann", 31}, {"Dwight", 19},
                                                                             {"Kassandra" , 21 }, {"Lawrence", 25 }, {"Cindy" , 22},
                                                                             {"Cory" , 27 }, {"Mac" , 19 }, {"Romana" , 27},
                                                                             {"Doretha" , 32 }, { "Danna" , 20}, {"Zara" , 23},
                                                                             {"Rosalyn" , 26 }, { "Risa" , 24}, { "Benny" , 28},
                                                                             { "Juan" , 33 }, {"Natalie" , 25 } };
            //display and order list of numbers
            Console.WriteLine("List of numbers:");
            display_num(numbers);
            Console.WriteLine("\nOrdered numbers:");
            var ordered_num = numbers.OrderBy(x => x).ToList();     //ordered numbers in ascending fashion
            display_num(ordered_num);
            //display and order list of people by name then age
            Console.WriteLine("List of people:");
            display_people(people);
            Console.WriteLine("\nOrdered people by name (alphabetically):");
            var alpha_ordered_people = people.OrderBy(p => p.Key).ToDictionary(p => p.Key, p => p.Value);   //ordered people alphabetically
            display_people(alpha_ordered_people);
            Console.WriteLine("\nOrdered people by descending age (then alphabetically):");
            var desc_age_ordered_people = alpha_ordered_people.OrderByDescending(p => p.Value).ToDictionary(p => p.Key, p => p.Value);  //ordered people descending by age then alphabetically
            display_people(desc_age_ordered_people);
            Console.WriteLine();
        }
    }
}
